File: lt_piece_rate_pay_report/lt_piece_rate_pay_report/report/production_status_details/production_status_details.js
// Copyright (c) 2026, LTL and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Production Status Details"] = {
	filters: [
		{
			fieldname: "po_list",
			label: __("PO List"),
			fieldtype: "Link",
			options: "PO List",
			reqd: 1,

			on_change: function () {
				frappe.query_report.set_filter_value("style_list", "");
				frappe.query_report.set_filter_value("color", "");
				load_styles();   // ✅ FIX 1
			}
		},
		{
			fieldname: "style_list",
			label: __("Style"),
			fieldtype: "Select",
			reqd: 1,
			on_change: function () {
				frappe.query_report.set_filter_value("color", "");
				load_colors();
			}
		},
		{
			fieldname: "color",
			label: __("Color"),
			fieldtype: "Select"
		},
		{
			"fieldname": "process_type",
			"fieldtype": "Link",
			"label": "Process Type",
			"mandatory": 0,
			"options": "Process Type",
			"wildcard_filter": 0
		},
		{
			"fieldname": "floor",
			"fieldtype": "Link",
			"label": "Floor",
			"mandatory": 0,
			"options": "Floor",
			"wildcard_filter": 0
		},
        {
			"fieldname": "line",
			"fieldtype": "Link",
			"label": "Line",
			"mandatory": 0,
			"options": "Facility or Line",
			"wildcard_filter": 0
		},
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
		},
	]
};

// ---------------- FUNCTIONS ----------------

function load_styles() {
	let po = frappe.query_report.get_filter_value("po_list");
	if (!po) return;

	frappe.call({
		method: "lt_piece_rate_pay_report.lt_piece_rate_pay_report.report.production_status_details.production_status_details.get_styles",
		args: { po_list: po },
		callback: function (r) {
			frappe.query_report.get_filter("style_list").df.options =
				["", ...(r.message || [])];
			frappe.query_report.get_filter("style_list").refresh();
		}
	});
}

function load_colors() {
	let po = frappe.query_report.get_filter_value("po_list");
	let style = frappe.query_report.get_filter_value("style_list");
	if (!po || !style) return;

	frappe.call({
		method: "lt_piece_rate_pay_report.lt_piece_rate_pay_report.report.production_status_details.production_status_details.get_colors",
		args: {
			po_list: po,
			style: style
		},
		callback: function (r) {
			frappe.query_report.get_filter("color").df.options =
				["", ...(r.message || [])];
			frappe.query_report.get_filter("color").refresh();
		}
	});
}

File: lt_piece_rate_pay_report/lt_piece_rate_pay_report/report/production_status_details/production_status_details.py
# Copyright (c) 2025, Lithe-Tech LTD and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):

	if not filters:
		filters = {}

	columns = get_columns()
	data = get_data(filters)
	
	return columns, data

def get_columns():
	return [
        _("Production Date") + ":Data:130",
        _("Color") + ":Data:150",
        _("Process Type") + ":Data:100",
		_("Floor") + ":Data:70",
		_("Line") + ":Data:50",
        # _("Style") + ":Data:90",
        # _("Sales Contract") + ":Data:200",
        _("Total Qty") + ":Int:100",
        _("Completed Qty") + ":Int:100",
		_("Bill Qty") + ":Int:100",
		_("Remain Qty") + ":Int:100",

        # _("Stamp Deduct") + ":Int:100",
		# _("Total Bill") + ":Int:100",

    ]

def get_data(filters):

	conditions, filters = get_conditions(filters)


	result = frappe.db.sql("""
        SELECT 
            dp.production_date,
            dp.color,
            dp.process_type,
			dp.floor,
			dp.facility_or_line,
            dp.total_quantity,
            dp.completed_quantity,
            dp.bill_quantity,
			dp.total_quantity-dp.completed_quantity-dp.bill_quantity
            
        FROM `tabDaily Production` dp
        WHERE %s
		ORDER BY FIELD(process_type, 'cutting', 'sewing', 'iron');
    """ % conditions, as_list=1)

	return result

def get_conditions(filters):
	conditions="1=1" 
	if filters.get("po_list"):conditions += " AND dp.po = '%s'" % filters["po_list"]
	if filters.get("style_list"):conditions += " AND dp.style_list = '%s'" % filters["style_list"]
	if filters.get("color"):conditions += " AND dp.color LIKE CONCAT('%%', {0}, '%%')".format(
		frappe.db.escape(filters["color"])
	)
	if filters.get("process_type"):conditions += " AND dp.process_type = '%s'" % filters["process_type"]
	if filters.get("floor"):conditions += " AND dp.floor = '%s'" % filters["floor"]
	if filters.get("line"):conditions += " AND dp.facility_or_line = '%s'" % filters["line"]
	if filters.get("company"): conditions += " AND dp.company = '%s'" % filters["company"]
	return conditions, filters



@frappe.whitelist()
def get_styles(po_list):
    po = frappe.get_doc("PO List", po_list)

    styles = set()
    for row in po.po_details:
        if row.style:
            styles.add(row.style)

    return list(styles)


@frappe.whitelist()
def get_colors(po_list, style):
    po = frappe.get_doc("PO List", po_list)

    colors = set()
    for row in po.po_details:
        if row.style == style and row.color:
            colors.add(row.color)

    return list(colors)
